### ISSUE TITLE
AWS Explorer missing information when querying DynamoDB

### ISSUE THREAD

**Issue by rboy1** [aws/aws-toolkit-visual-studio#121]
**Describe the bug**
There's a serious bug in the AWS Explorer add on for Visual Studio 2019.

When Scanning a table with a specific condition it returns no/incomplete results in AWS Explorer. However when running the Scan through the browser console or an API it returns the correct list of results.

It appears to be out of sync, for example I see 5 entries returned by the web console in one Scan, where as AWS explorer returns only 4.
For another table the web console returns 1 entry where as AWS explorer returns none!

Example:
When I use the AWS Explorer from Visual Studio 2019, it returns only 4 objects (it's missing an entry from yesterday): https://drive.google.com/file/d/1iBcXtjLcQwZ7qbYFyaHFvkHVaACu2dcT/view?usp=sharing
When I use the AWS Web Console, it returns 5 objects (the correct number): https://drive.google.com/file/d/1_eas_PHq66Xlu95qqcDwF3Iq30aBpPBm/view?usp=sharing

This doesn't happen for every attribute value being matched by the Scan. What I mean is, if I change the value being searched by the "Scan" filters (in this screenshot the value is xxx975), for some values it retrieves all the records in AWS Explorer, but for some values (like the one in this screenshot xxx975), it's missing one record in the response.

BTW, this problem only happens with AWS Explorer. When using the same "Scan" via an API (in a .NET application) or through the Web Console it returns all the matching objects. It almost sounds like the results are not being completely enumerated by AWS Explorer after the Scan is completed and consequently it's missing a few records.

More details provided here: Screenshot are available here: https://forums.aws.amazon.com/thread.jspa?messageID=956512

**Expected behavior**
Return all matching items as returned by the Web Console

**Screenshots**
Screenshot are available here: https://forums.aws.amazon.com/thread.jspa?messageID=956512

**Computer (please complete the following information):**

- Windows Version: Windows 10
- Visual Studio Version: 1.18.1.0
- AWS Toolkit for Visual Studio Version: 16.7.3

**Additional context**

**Comment by awschristou** [comment]
Hi @rboy1 thanks for reaching out. I created a test table with ten items and three attributes (2 strings, 1 number). A "string equals x" scan returned the same results in the Toolkit and web console. Is there a chance that your toolkit scan is retrieving data before it is read consistent?

If you are able to create a small sample data set using placeholder data to illustrate the issue, it could help diagnose the problem.

**Comment by rboy1** [comment]
Yes that sounds about right. However it’s not happening always and I also found that this happening for new entries in the table. For example if I run the same scan after 24 hours it returns all objects. Only with specific scans show this issue like shown in the images. There’s a bug in the aws explorer code, one guess is that it should probably enumerated the all the objects before displaying if there is less than one page of items (Rows) returns.

**Comment by rboy1** [comment]
And to note, it's only missing the last object in these scans that's missing in AWS Explorer (especially if that entry was updated or created within the last 24 hours).

**Comment by awschristou** [comment]
In your screen shot it looks like there are buttons in the lower left to load more result pages. Do these provide the expected result set?

![image](https://user-images.githubusercontent.com/39839589/93239894-440aa900-f738-11ea-8322-111428a8a6ff.png)

**Comment by rboy1** [comment]
There are only 5 result entries and AWS Explorer is only showing 4. See the second screenshot from the Web console which shows the 5 results.

If you're saying did I try clicking that button to see if the last missing items shows up, yes I did but it didn't make any difference.

**Comment by awschristou** [comment]
Do you have a rough idea of how many items are in the table?

If the toolkit has scanned through the entire table, those buttons should become disabled. Have you clicked those buttons to the point where they become disabled? If your table has thousands of items, I'm thinking it is possible that the scan operation is going through some pages of items, and not finding any filtered results. This means you'd have to keep requesting additional pages until your 5th result appears.

**Comment by rboy1** [comment]
I think you've got the bug. I'm able to replicate the issue here with another scan which is supposed to return 32 results, but only returns 23. Then I click on the next page a couple of times and then finally all 32 show up.

The table has a total of 5K entries and it's expected to continue growing to many times that. Are you saying this is a bug with AWS Explorer or DynamoDB can't handle that many entries in a table or that the same issue will happen everywhere (including API's and web console)?

**Comment by awschristou** [comment]
Nice, sounds like we're getting closer to understanding what's going on. My suspicion is that when you use the API directly, and you paginate through the scan results, your code continues processing the pages until the end is reached. Maybe the web console is doing something similar. The Toolkit is simpler in concept, where those buttons currently perform a single "next page" request, regardless of whether the page contained contents.

Sounds like there is room for some UX improvement. There's a balance to be struck however - proceeding through subsequent pages too quickly could exceed the table's read capacity.

**Comment by rboy1** [comment]
With the .NET API's they just iterate the results; just so I understand the expected behavior, that would mean DynamoDB is automatically scanning the entire table - is that correct? If not, I need to know because it's coded with the expectation that iterating the results will scan everything (like a traditional DB).

Coming back to AWS Explorer:
If the table is set to "dynamic", my understanding is that it doesn't have any set read capacity. Is that correct?

If so can the tool not check the current capacity setting and if it's set to dynamic, just scan the entire table automatically OR
IMHO, it should scan the whole table until the results are atleast one page. This would make it behave more consistently with the UI expectation, i.e. for me the next button indicates that there's more than one page of "results"

**Comment by awschristou** [comment]
I'm under the impression that auto-scaling tables can still hit a read capacity limit if enough inbound requests reach a table before the read capacity is adjusted. If the toolkit were to try to automatically read to the end, it could get throttled,

Some UI thoughts I've had so far:
* the "load next page(s)" buttons might be able to surface information about how much of the table has been scanned (reducing the illusion that it didn't do anything)
* another button could be added: "make your best attempt at scanning the rest of the table", which could stop if throttled, and could be pressed again to continue scanning

**Comment by rboy1** [comment]
Okay I would like to split this discussion into 3 sub section if you're okay:
1. Does this issue of "scanning" hitting a read capacity limit also affect a "query"? The documents are very clear on the difference. If not, then couldn't AWS Explorer use a "query" instead of a "scan"?
2. How does it impact the API (I'm trying to understand this as it will also potentially impact my .NET API implementation), i.e. if the app continues to iterate the result object, will it automatically trigger DynamoDB to "read" the next set of pages or does the API need to make a separate call to "continue" reading?
3. Regarding the UI, assuming that 1 above cannot resolve the issue, it may be better to have a check box which says continue reading until all results are returned. This also helps with another issue I noticed earlier, I wanted to "dump" the entire table into an excel sheet and I just couldn't find any way to do it from AWS Explorer. With this option it'll allow me to do that.

**Comment by normj** [comment]
Throwing in my opinions.

1. It is true the toolkit could be smarter and based on the query configured in the toolkit and if it the query is compatible with DDB query operation we could switch to query mode. That would mean if the toolkit is configured to search on an exact hash key and a filter on query so its usefulness would be limited.

2. Will have to be very careful with a display all results button. Your table is relatively small which would result in about 5 pagination calls. Imagine the DynamoDB tables that Amazon.com uses and the billions and trillions of records those tables contain. I do agree better UI up on top of the table browser that alerts the user that more data might exists would make it clearer. Also maybe doing a few pages of data up to a X number of seconds instead of just one page might be better. Then let the user know this is much data as we found in X seconds click the button if you want to the toolkit to search for X more seconds.

3. The toolkit is using the .NET SDK to perform the operation. If you are using the service client, the AmazonDynamoDBv2 client then you had to code to handle the scan pagination. The DataModel and Document high level abstractions in the SDK do make the pagination transparent under the covers. Also if you are using the new pagination feature that was released last month in the SDK that can make the experience of handling pagination transparent. Regardless of either approach they will consume read capacity units for each page regardless of you manually paginating or having it done for you.

**Comment by rboy1** [comment]
@normj I like your option 2, that's better than just stopping at one page. Coupled with an explicit pop/notification that the results may be incomplete IF the results are less than one "display" page. If it's more than one "display" page the user will automatically look at clicking the "Next" page button anyways so I don't see an issue there. I would also love to see a "100" pagination option if I'm trying to export my entire table.

Regarding 3, I'm using AmazonDynamoDBv2 which was implemented a few months ago, how do I know if the pagination feature is being used or if I need to manually check? As in, could you point me to "updated" docs or the SDK version number that I can check to see I'm handling it correctly. I'm assuming this applies to both Query and Scans.
